pkg/rpc/context.go

```diff
@@ -2204,18 +2264,90 @@ func (rpcCtx *Context) grpcDialNodeInternal(
 ) *Connection {
 	k := connKey{target, remoteNodeID, class}
 	if conn, ok := rpcCtx.m.Get(k); ok {
-		// There's a cached connection.
+		// Returns cached connection that isn't neither failed nor decommissioned.
 		return conn
 	}
-
 	ctx := rpcCtx.makeDialCtx(target, remoteNodeID, class)
 
-	conn, inserted := rpcCtx.m.TryInsert(k)
+	conn, pc, inserted := rpcCtx.m.TryInsert(k)
 	if !inserted {
 		// Someone else won the race.
 		return conn
 	}
 
+	// TODO (koorosh): breaker is stored per `peer` and shouldn't be reinitialized every time `grpcDialNodeInternal`
+	// is called.
+	if pc.breaker == nil {
+		// TODO (koorosh): Probably this logic should be extracted from here...
+		breaker := circuitbreaker.NewBreaker(circuitbreaker.Options{
+			Name: "reconnect",
+			AsyncProbe: func(report func(error), done func()) {
+				if err := rpcCtx.Stopper.RunAsyncTask(ctx, "reconnect peers", func(ctx context.Context) {
+					var t timeutil.Timer
+					defer t.Stop()
+					defer done()
+					for {
+						// periodically run probe until connection is healed.
+						// TODO (koorosh): would it be enough to pause for `heartbeatInterval` duration?
+						t.Reset(rpcCtx.heartbeatInterval)
+						select {
+						case <-rpcCtx.Stopper.ShouldQuiesce():
+							return
+						case <-t.C:
+							t.Read = true
+						}
+						pc.mu.RLock()
+						disconnected := pc.mu.disconnected
+						decommissioned := pc.mu.decommissioned
+						pc.mu.RUnlock()
+						if disconnected.IsZero() || decommissioned {
+							report(nil)
+							return
+						}
+						c, _, inserted := rpcCtx.m.TryInsert(k)
+						if !inserted {
+							// someone else inserted connection and probably
+							// started dialing.
+							continue
```

Comment:
Why does it matter who inserted? We are just trying to make sure there's a Connection, and we want to see it in a healthy state (call `Connect`). Why the `continue` here?